File: baekjoon/cpp/level18/9012.cpp
#include<iostream>
#include<stack>

using namespace std;

int main(){
    int n;
    cin >> n;

    string str;
    for(int i=0;i<n;i++){
        bool flag = false;
        stack<char> st;
        cin >> str;
        int len = str.length();
            for(int j=0; j<len; j++){
                if(str[j] == '('){
                    st.push(str[j]);
                }
                else if(str[j] == ')'){
                    if(st.size() > 0){
                        if(st.top() == '('){
                            st.pop();
                        }
                        else{
                            break;
                        }
                    }
                    else{
                        st.push(str[j]);
                        break;
                    }
                }
            }
            if(st.size() == 0) cout << "YES\n";
            else cout << "NO\n";
    }
}

File: baekjoon/js/level20 Stack/9012.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "../input.txt";
let [N,...input] = fs.readFileSync(filePath).toString().trim().split("\n");

const ans = [];

input.forEach((elem) => {
  const item = elem.split('');
  const stack = [];
  for(let i=0; i<item.length; i++){
    if(item[i] === '(')
      stack.push('(');
    else{
      if(0 < stack.length){
          stack.pop();
      }
      else{
        stack.push(')'); break;
      }
    }
  }
  stack.length ? ans.push('NO') : ans.push('YES');
})
console.log(ans.join('\n'));